### PR TITLE
Housekeeping/code cleanup pre commit

### PR DESCRIPTION
**Pull request by mariusconjeaud**
- Replaced RelationshipTo / RelationshipFrom / Relationship methods with classes for consistency (Property and its descendants are classes). This is not a breaking change.
- Added pre-commit hooks (black for prettifying and isort for imports)
- Cleaned up part of the code with the help of pylint (ultimate goal is to add pylint as pre-commit / pipeline check)

### CODE REVIEW

File: neomodel/relationship.py [hunk 1]
```diff
@@ -1,32 +1,36 @@
...
 
 class StructuredRel(StructuredRelBase):
     """
     Base class for relationship objects
     """
+
     def __init__(self, *args, **kwargs):
-        super(StructuredRel, self).__init__(*args, **kwargs)
+        super().__init__(*args, **kwargs)
+        self._start_node_id = 0
```
Comment (mariusconjeaud) [inline neomodel/relationship.py:31]: pylint issue fix (member properties not set in init)

File: neomodel/relationship.py [hunk 2]
```diff
@@ -37,40 +41,48 @@ def save(self):
         """
         props = self.deflate(self.__properties__)
         query = "MATCH ()-[r]->() WHERE id(r)=$self "
-        for key in props:
-            query += " SET r.{0} = ${1}".format(key, key)
-        props['self'] = self.id
+        query += "".join([f" SET r.{key} = ${key}" for key in props])
```
Comment (mariusconjeaud) [inline neomodel/relationship.py:44]: improved syntax

File: neomodel/relationship_manager.py [hunk 3]
```diff
@@ -193,31 +226,35 @@ def reconnect(self, old_node, new_node):
...
+            "MATCH (us), (old), (new) "
+            "WHERE id(us)=$self and id(old)=$old and id(new)=$new "
             "MATCH " + old_rel
+        )
         q += " MERGE" + new_rel
 
         # copy over properties if we have
         for p in existing_properties:
-            q += " SET r2.{0} = r.{1}".format(p, p)
+            q += "".join([f" SET r2.{prop} = r.{prop}" for prop in existing_properties])
```
Comment (mariusconjeaud) [inline neomodel/relationship_manager.py:254]: improved syntax

File: neomodel/relationship_manager.py [hunk 4]
```diff
@@ -361,61 +400,78 @@ def __getitem__(self, key):
...
+class RelationshipDefinition:
+    def __init__(
+        self,
+        relation_type,
+        cls_name,
+        direction,
+        manager=RelationshipManager,
+        model=None,
+    ):
+        self._validate_class(cls_name, model)
```
Comment (mariusconjeaud) [inline neomodel/relationship_manager.py:412]: Moved up validation into parent class (see updates to Relationship/From below)

File: neomodel/relationship_manager.py [hunk 5]
```diff
@@ -361,61 +400,78 @@ def __getitem__(self, key):
...
-        def enumerate_traceback(initial_frame):
-            depth, frame = 0, initial_frame
-            while frame is not None:
-                yield depth, frame
-                frame = frame.f_back
-                depth += 1
+        current_frame = inspect.currentframe()
 
-        frame_number = 4
+        frame_number = 3
```
Comment (mariusconjeaud) [inline neomodel/relationship_manager.py:416]: Trace is one level smaller (see updates to RelationshipTo/From below)

File: neomodel/relationship_manager.py [hunk 6]
```diff
@@ -361,61 +400,78 @@ def __getitem__(self, key):
...
+                        raise RelationshipClassRedefined(
+                            relation_type, db._NODE_CLASS_REGISTRY, model
+                        )
                 else:
                     db._NODE_CLASS_REGISTRY[label_set] = model
             except KeyError:
                 # If the mapping does not exist then it is simply created.
                 db._NODE_CLASS_REGISTRY[label_set] = model
 
+    def _validate_class(self, cls_name, model):
```
Comment (mariusconjeaud) [inline neomodel/relationship_manager.py:459]: Extracted validation in separate method

File: neomodel/relationship_manager.py [hunk 7]
```diff
@@ -453,25 +509,44 @@ class ZeroOrMore(RelationshipManager):
     """
     A relationship of zero or more nodes (the default)
     """
-    description = "zero or more relationships"
-
-
-def _relate(cls_name, direction, rel_type, cardinality=None, model=None):
```
Comment (mariusconjeaud) [inline neomodel/relationship_manager.py:459]: Refactored RelationshipTo, RelationshipFrom, Relationship from methods to classes for consistency (Property objects are classes, not methods).
This is not a breaking change.

File: pyproject.toml [hunk 8]
```diff
@@ -48,14 +48,31 @@ changelog = "https://github.com/neo4j-contrib/neomodel/releases"
 [project.optional-dependencies]
 dev = [
     "pytest>=7.1",
+    "pytest-cov>=4.0",
```
Comment (mariusconjeaud) [inline pyproject.toml:51]: Added dependencies for pre-commit actions

File: pyproject.toml [hunk 9]
```diff
@@ -48,14 +48,31 @@ changelog = "https://github.com/neo4j-contrib/neomodel/releases"
...
     "pre-commit",
+    "black",
+    "isort",
     "Shapely>=1.8.1,<1.9"
 ]
 
 [tool.pytest.ini_options]
-addopts = "--resetdb"
+addopts = "--resetdb --cov=neomodel --cov-report=html:coverage_report"
```
Comment (mariusconjeaud) [inline pyproject.toml:59]: Add coverage to test process

File: pyproject.toml [hunk 10]
```diff
@@ -48,14 +48,31 @@ changelog = "https://github.com/neo4j-contrib/neomodel/releases"
...
 testpaths = "test"
 
+[tool.isort]
+profile = 'black'
+src_paths = ['neomodel']
+
+[tool.pylint.'MESSAGES CONTROL']
+disable = 'missing-module-docstring,redefined-builtin,missing-class-docstring,missing-function-docstring,consider-using-f-string,line-too-long'
```
Comment (mariusconjeaud) [inline pyproject.toml:67]: Add pylint ignores ; some are temporary to fix issues step by step (like f-string or line-too-long)

**Review by mariusconjeaud — COMMENTED** [review]
Marking the files to look at (to ignore changes made only by black)

File: neomodel/relationship_manager.py [hunk 11]
```diff
@@ -453,25 +509,44 @@ class ZeroOrMore(RelationshipManager):
...
-
-
-def RelationshipFrom(cls_name, rel_type, cardinality=ZeroOrMore, model=None):
-    return _relate(cls_name, INCOMING, rel_type, cardinality, model)
+    description = "zero or more relationships"
 
 
-def Relationship(cls_name, rel_type, cardinality=ZeroOrMore, model=None):
-    return _relate(cls_name, EITHER, rel_type, cardinality, model)
+class RelationshipTo(RelationshipDefinition):
```
Comment (aanastasiou) [inline neomodel/relationship_manager.py:516]: This is probably slightly faster, during definition but the two are equivalent. 
Perhaps more importantly: Let's start documenting these classes, it is probably the number one cause of why there are diversions between the docs and the code.

**Review by aanastasiou — APPROVED** [review]
The effect of Black has been a bit confusing, it would probably have been better to have the review before it run on the code.

I specifically focused on the Relationship change and the deterministic indices, which are fine.

**Review by tonioo — APPROVED** [review]

**Pull request merged**